Sources/Compiler/AST/Expr/FoldedSequenceExpr.swift

```diff
@@ -0,0 +1,66 @@
+/// A sequence of binary operations stored as a tree whose structure encodes the evaluation order.
+///
+/// Instances of this type are created during type checking for each `SequenceExpr` in the program
+/// once the precedence groups of its operators have been determined. A tree is created by calling
+/// `append(operator:right:)` to append an operator and its right operand to the sub-sequence
+/// represented by `self`.
+public indirect enum FoldedSequenceExpr {
+
+  /// The expression of an operator in the AST together with its precedence.
+  public typealias Operator = (expr: NodeID<NameExpr>, precedence: PrecedenceGroup?)
+
+  /// A parent node representing the application of `operator` on `left` and `right`.
+  case parent(operator: Operator, left: FoldedSequenceExpr, right: FoldedSequenceExpr)
+
+  /// A leaf node representing some expression.
```

Comment:
```suggestion
  /// An expression that is not, at the top level, an operator application.
```

Or something.  What you wrote is obvious from the declaration.  Use the opportunity provided by the comment to say something more.

Reply:
Does the precision adds anything useful to the reader?

What you wrote is wrong because `AnyExprID` can represent, well any expression. That includes another `SequenceExpr`, which doesn't violate well-formedness if it was obtained from`a * (b + c)`. Further `!a` is also an operator application that certainly can be a leaf node.

I feel like adding qualification to exclude this cases will just make the comment heavier.

Reply:
> Does the precision adds anything useful to the reader?

Absolutely, yes.

> What you wrote is wrong

Which goes to show that I don't understand the abstraction being documented here.